### PR TITLE
WIP: Add missing trait implementation lints

### PR DESCRIPTION
**Pull request by Susurrus**
Posting this now so I can ensure I'm on the right track here.

This adds lints for if types are missing implementations of:

- [ ] `Clone`
- [ ] `Copy`
- [ ] `Debug`
- [ ] `Default`
- [ ] `Display`
- [ ] `Eq`
- [ ] `Hash`
- [ ] `Ord`
- [ ] `PartialEq`
- [ ] `PartialOrd`

Context here is in implementing [RFC2235](https://github.com/rust-lang/rfcs/blob/master/text/2235-libc-struct-traits.md) we cannot auto-derive all traits. However, we still want the traits to be manually implemented. So we'd like lints to check on them.

I started this work in rust-lang/rust#58070, but it was suggested that these lints shouldn't be in the compiler and instead in Clippy. It was suggested that the existing lints for `Copy` and `Debug` be deprecated as part of this as well.

Additionally these lints are part of #1798.

To fully implement all lints here, rust-lang/rust#58070 will need to be merged (which is still a work in progress) in order to get `Default`, `Display`, and `Hash` added as lang_items so they can be easily checked on types.

### CODE REVIEW

**Comment by mikerite** [comment]
`correctness` is the wrong category for these lints. The lints in the `correctness` category is for lints that detect bugs. This probably belongs in the `pedantic` category.

Also it looks like there is no code for checking if it's even possible to implement the traits so this might have a lot of false positives.

**Comment by Susurrus** [comment]
> `correctness` is the wrong category for these lints. The lints in the `correctness` category is for lints that detect bugs. This probably belongs in the `pedantic` category.

Where do I have them set as part of the `correctness` category? I believe I added them to the `clippy::all` group and that's it. However, I don't understand the lint code very well, so it's possible I missed something here. I wasn't certain if these lints belonged in any of those categories, which is why I tried to just add them to `clippy::all`.

> Also it looks like there is no code for checking if it's even possible to implement the traits so this might have a lot of false positives.

This came up during the discussion in my Rust PR. I believe the lint for `Copy` checked whether it could be implemented. I don't, however, believe this applies to all of these. For example `Clone`, `Debug`, `Default`, `Display`, `Hash`, and `PartialEq` I would expect to be implementable for all types. `Copy`, `Eq`, `Ord`, and `PartialOrd` I'd expect to have false positives.

File: clippy_lints/src/missing_traits.rs [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+use rustc::lint::{LateContext, LateLintPass, LintArray, LintPass};
+use rustc::{declare_tool_lint, lint_array};
+use rustc::util::nodemap::NodeSet;
+
+macro_rules! missing_impl {
+    ($trait:ident, $trait_path:path, $lint_constant:ident, $struct_name:ident,
+        $trait_check:ident) => (
+        declare_clippy_lint! {
+            pub $lint_constant,
+            correctness,
```
Comment (oli-obk) [inline clippy_lints/src/missing_traits.rs]: This is where you give a lint its category

**Comment by oli-obk** [comment]
> For example `Clone`, `Debug`, `Default`, `Display`, `Hash`, and `PartialEq` I would expect to be implementable for all types

You don't want `Clone` and `Default` for types where you should only be getting values from other values. One example is `MutexGuard`

**Comment by flip1995** [comment]
Ping from triage @Susurrus. Since rust-lang/rust#58070 is `S-inactive-closed`, and there's no update on this in a while either, I'll also set this to `S-inactive-closed`. Thanks for your contribution and feel free to reopen this any time!

**Pull request closed without merging**